src/with_alloc.rs

```diff
@@ -36,47 +64,57 @@ use core::mem::MaybeUninit;
 /// assert_eq!(buffer.to_vec(), vec![42, 1]);
 /// ```
 #[derive(Debug)]
-pub struct AllocRingBuffer<T> {
+pub struct AllocRingBuffer<T, MODE: RingbufferMode = PowerOfTwo> {
     buf: Vec<MaybeUninit<T>>,
     capacity: usize,
     readptr: usize,
     writeptr: usize,
+    mode: PhantomData<MODE>,
 }
 
-impl<T> Drop for AllocRingBuffer<T> {
+impl<T, MODE: RingbufferMode> Drop for AllocRingBuffer<T, MODE> {
     fn drop(&mut self) {
         self.drain().for_each(drop);
     }
 }
 
-impl<T: Clone> Clone for AllocRingBuffer<T> {
+impl<T: Clone> Clone for AllocRingBuffer<T, PowerOfTwo> {
     fn clone(&self) -> Self {
         let mut new = Self::with_capacity(self.capacity);
         self.iter().cloned().for_each(|i| new.push(i));
         new
     }
 }
-impl<T: PartialEq> PartialEq for AllocRingBuffer<T> {
+
+impl<T: Clone> Clone for AllocRingBuffer<T, NonPowerOfTwo> {
+    fn clone(&self) -> Self {
+        let mut new = Self::with_capacity_non_power_of_two(self.capacity);
+        self.iter().cloned().for_each(|i| new.push(i));
+        new
+    }
+}
```

Comment:
If I understand correctly, a function generic on `MODE: RingBufferMode` cannot use `<AllocRingBuffer<T, MODE> as Clone>` because the compiler doesn't understand the `Sealed` trait pattern.  So any users of `AllocRingBuffer` that need features like these will have to duplicate code for `PowerOfTwo` and `NonPowerOfTwo`.  It might be easier, thus, to perform a runtime branch on `MODE` within `clone()`; after all, monomorphization should constant-fold it.

Reply:
oh good catch, let's see if I can make this a bit better

Reply:
I don't think the runtime check is necessary even. We just use the unsafe version, taking the same capacity as the buffer we're cloning. We know that the old capacity is valid, so then the new one must be too (although I did add some debug asserts for it)

Reply:
alright, perfect